sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -609,6 +610,19 @@ object CollapseWindow extends Rule[LogicalPlan] {
 }
 
 /**
+ * Flip Adjacent Window Expressions.
+ * - If the partition spec of the parent Window expression is a subset of the partition spec
+ *   of the child window expression, flip them.
+ */
+object FlipWindow extends Rule[LogicalPlan] {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transformUp {
+    case w1 @ Window(we1, ps1, os1, w2 @ Window(we2, ps2, os2, grandChild))
+        if ps2.containsSlice(ps1) =>
```

Comment:
This condition might not be enough. `we1` might depend on the outputs of `we2`, right?